src/licensedcode/plugin_license.py

```diff
@@ -116,3 +118,57 @@ def get_scanner(
             license_text_diagnostics=license_text_diagnostics,
             license_url_template=license_url_template
         )
+    def process_codebase(self, codebase, **kwargs):
+
+        for resource in codebase.walk(topdown=False):
+            match_reference_license(resource,codebase)
+
+
+def match_reference_license(resource, codebase):
+    """
+    Find instances for any licenses in referenced filenames
+    """
+    licenses = resource.licenses
+    license_expressions = resource.license_expressions
+    if not licenses:
+        return 
+
+    location = resource.location
+    
+    from licensedcode import cache
+    from scancode.api import get_licenses
+    idx = cache.get_index()
+    matches = idx.match(
+        location=location, min_score=0)
+    
+    modified = False
+
+    for match in matches:
+        ref_files=match.rule.referenced_filenames
+        if len(ref_files) != 0:
+            for i in range(len(ref_files)):
+                if not ref_files[i].startswith('usr/share/common-licenses'):
+                 new_loc=find_reference_file(location,ref_files[i])
+                 if new_loc != None:
+                   new_lic=get_licenses(new_loc, min_score=0)
```

Comment:
> does process_codebase functions run after all scanners have completed running, i.e. when process_codebase is called for a resource for a scan plugin, have all resources already been populated with LicenseMatch objects detected in them? (The above is assuming that)

- For a scanner plugin, a `process_codebase` function is guaranteed to be called after this scanner has completed for all files. (it may be that all scanners have been running for that file, but that's not guaranteed)
- For a post-scan plugin, a `process_codebase` function is guaranteed to be called after all the scanners have completed for all files.

Reply:
So for the license scanner, yes, it has finished running on all files when its `process_codebase` is called.

Reply:
This makes sense. Thanks @pombredanne

Reply:
> This is not the correct way to do this.
> 
> Currently you are:
> 
> 1. Finding the directory locally with the location.
> 2. Doing a license detection on the file that you have found.
> 
> The approach should have been, as mentioned [here](https://github.com/nexB/scancode-toolkit/issues/1364#issuecomment-888988049) to
> 
> 1. get the parent resource from the current resource
> 2. check if the referenced filename is present in the child resources of this parent resource.
> 3. if resource object with referenced filename present, get the licenses from the from the resource object.
> 4. then add the matches and license expression.

The parent that we get from Resource object is also a Resource object right? When I try to get children from parent object(using resource.parent) it throws attribute error altho if if parent is also a Resource objecr we can use this https://github.com/nexB/commoncode/blob/main/src/commoncode/resource.py#L1188 to get children

Reply:
It could be an iterable (?). You should be able to for loop through the children like [this](https://github.com/nexB/commoncode/blob/main/src/commoncode/resource.py#L1171)

Reply:
Yup tried this only ,but it is showing attribute error

Reply:
Could you paste the error here please. Also try these checking functions 1. https://github.com/nexB/commoncode/blob/main/src/commoncode/resource.py#L1211 2. https://github.com/nexB/commoncode/blob/main/src/commoncode/resource.py#L1199 3. https://github.com/nexB/commoncode/blob/main/src/commoncode/resource.py#L1182

Reply:
plugin.process_codebase(codebase, **kwargs)
  File "/home/akanksha/Desktop/git/scancode-toolkit/src/licensedcode/plugin_license.py", line 129, in process_codebase
    match_reference_license(resource,codebase)
  File "/home/akanksha/Desktop/git/scancode-toolkit/src/licensedcode/plugin_license.py", line 153, in match_reference_license
    child = par.children
AttributeError: 'function' object has no attribute 'children'

Reply:
Same attribute error for all all these functions . One point is if we try to access these func from resource directly it works! Ex: _for child in resource.children(codebase)_this will work perfectly.
Seems like parent is not returning a Resource object ???

Reply:
Ah. `resource.parent` is a function, not the `Resource` object. `resource.parent(codebase)` actually returns the parent Resource object you need. :P

Reply:
oops ! right silly mistake ;(